libfreerdp/core/rdstls.c

```diff
@@ -0,0 +1,784 @@
+
+	if (!Stream_EnsureRemainingCapacity(s, length))
+		return FALSE;
+
+	Stream_Write(s, data, length);
+
+	return TRUE;
+}
+
+static BOOL rdstls_write_authentication_request_with_password(rdpRdstls* rdstls, wStream* s)
+{
+	rdpSettings* settings = rdstls->context->settings;
+	WINPR_ASSERT(settings);
+
+	if (!Stream_EnsureRemainingCapacity(s, 4))
+		return FALSE;
+
+	Stream_Write_UINT16(s, RDSTLS_TYPE_AUTHREQ);
+	Stream_Write_UINT16(s, RDSTLS_DATA_PASSWORD_CREDS);
+
+	if (!rdstls_write_data(s, settings->RedirectionGuidLength, settings->RedirectionGuid))
+		return FALSE;
+
+	if (rdstls_write_string(s, settings->Username) < 0)
+		return FALSE;
+
+	if (rdstls_write_string(s, settings->Domain) < 0)
+		return FALSE;
+
+	if (!rdstls_write_data(s, settings->RedirectionPasswordLength, settings->RedirectionPassword))
+		return FALSE;
+
+	return TRUE;
+}
+
+static BOOL rdstls_write_authentication_request_with_cookie(rdpRdstls* rdstls, wStream* s)
+{
+	// TODO
+	return FALSE;
+}
+
+static BOOL rdstls_write_authentication_response(rdpRdstls* rdstls, wStream* s)
+{
+	if (!Stream_EnsureRemainingCapacity(s, 8))
+		return FALSE;
+
+	Stream_Write_UINT16(s, RDSTLS_TYPE_AUTHRSP);
+	Stream_Write_UINT16(s, RDSTLS_DATA_RESULT_CODE);
+	Stream_Write_UINT32(s, rdstls->resultCode);
+
+	return TRUE;
+}
+
+static BOOL rdstls_process_capabilities(rdpRdstls* rdstls, wStream* s)
+{
+	UINT16 dataType;
+	UINT16 supportedVersions;
+
+	if (!Stream_CheckAndLogRequiredLength(TAG, s, 4))
+		return FALSE;
+
+	Stream_Read_UINT16(s, dataType);
+	if (dataType != RDSTLS_DATA_CAPABILITIES)
+	{
+		WLog_ERR(TAG, "received invalid DataType=0x%04" PRIX16 ", expected 0x%04" PRIX16, dataType,
+		         RDSTLS_DATA_CAPABILITIES);
+		return FALSE;
+	}
+
+	Stream_Read_UINT16(s, supportedVersions);
+	if ((supportedVersions & RDSTLS_VERSION_1) == 0)
+	{
+		WLog_ERR(TAG, "received invalid supportedVersions=0x%04" PRIX16 ", expected 0x%04" PRIX16,
+		         supportedVersions, RDSTLS_VERSION_1);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static BOOL rdstls_read_unicode_string(wStream* s, char** str)
+{
+	UINT16 length = 0;
+
+	WINPR_ASSERT(str);
+
+	if (!Stream_CheckAndLogRequiredLength(TAG, s, 2))
+		return FALSE;
+
+	Stream_Read_UINT16(s, length);
+
+	if (!Stream_CheckAndLogRequiredLength(TAG, s, length))
+		return FALSE;
+
+  if (length <= 2)
+  {
+    Stream_Seek(s, length);
+    return TRUE;
+  }
+
+	*str = Stream_Read_UTF16_String_As_UTF8(s, length / sizeof(WCHAR), NULL);
+	if (!*str)
+		return FALSE;
+
+	return TRUE;
+}
+
+static BOOL rdstls_read_data(wStream* s, UINT16* pLength, BYTE** pData)
+{
+	UINT16 length = 0;
+
+	WINPR_ASSERT(pLength);
+	WINPR_ASSERT(pData);
+
+	if (!Stream_CheckAndLogRequiredLength(TAG, s, 2))
+		return FALSE;
+
+	Stream_Read_UINT16(s, length);
+
+	if (!Stream_CheckAndLogRequiredLength(TAG, s, length))
+		return FALSE;
+
+	if (length <= 2)
+  {
+    Stream_Seek(s, length);
+    return TRUE;
+  }
+
+	free(*pData);
+	*pData = (BYTE*)malloc(length);
```

Comment:
no need to allocate, your only use just needs a pointer to the ` wStream` and a size, then you can remove the free in `rdstls_process_authentication_request_with_password`